File: day4/day4.py
# -*- coding: utf-8 -*-

### DAY 4

## PART 1

def elfRange(endpoints):
    enpoints = endpoints.split("-")
    enpoints = [eval(i) for i in enpoints]
    listRange = list(range(enpoints[0], enpoints[1]+1))
    return(listRange)

with open(r"data.txt", "r") as f:
    areContainedInOther = 0

    for line in f:
        line = line.strip().split(",")

        elfOne = elfRange(line[0])
        elfTwo = elfRange(line[1])

        overlaps = set(elfOne).intersection(elfTwo)

        if len(overlaps) == len(elfOne) or len(overlaps) == len(elfTwo):
            areContainedInOther += 1

print(f"In {areContainedInOther} pairs one range is fully contained in the other.")


## PART 2

with open(r"data.txt", "r") as f:
    areOverlapping = 0

    for line in f:
        line = line.strip().split(",")

        elfOne = elfRange(line[0])
        elfTwo = elfRange(line[1])

        overlaps = set(elfOne).intersection(elfTwo)

        if len(overlaps) > 0:
            areOverlapping += 1

print(f"In {areOverlapping} pairs the two ranges overlap.")

File: R_solutions/day4.R
library(data.table)

### DAY 4

args <- commandArgs(trailingOnly = T)
data <- args[which(args == "-in") + 1]

data <- fread(data, header = F, col.names = c("elf.one", "elf.two"))

## PART 1

range.endpoints <- function(range) {
  return(as.integer(strsplit(range, split = "-")[[1]]))
}

is.contained.in.other <- sapply(1:nrow(data), function(i) {
  
  nums <- range.endpoints(as.vector(data[i, elf.one]))
  e1 <- nums[1]:nums[2]
  
  nums <- range.endpoints(as.vector(data[i, elf.two]))
  e2 <- nums[1]:nums[2]
  
  overlap.length <- length(intersect(e1, e2))
  
  if(length(e1) == overlap.length || length(e2) == overlap.length) {
    return(T)
  } else {
    return(F)
  }
})

cat("In ",
    sum(is.contained.in.other),
    " pairs one range is fully contained in the other.\n",
    sep = "")


## PART 2

is.overlapping.other <- sapply(1:nrow(data), function(i) {
  
  nums <- range.endpoints(as.vector(data[i, elf.one]))
  e1 <- nums[1]:nums[2]
  
  nums <- range.endpoints(as.vector(data[i, elf.two]))
  e2 <- nums[1]:nums[2]
  
  overlap.length <- length(intersect(e1, e2))
  
  if(overlap.length > 0) {
    return(T)
  } else {
    return(F)
  }
})

cat("In ",
    sum(is.overlapping.other),
    " pairs the two ranges overlap.\n",
    sep = "")
